### DISCUSSION TITLE
8-ways-to-set-the-urls-for-an-aspnetcore-app/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#410] [Announcements]
# 8-ways-to-set-the-urls-for-an-aspnetcore-app/

In this post I describe 8 different ways to set which URLs your ASP.NET Core application listens on.

https://andrewlock.net/8-ways-to-set-the-urls-for-an-aspnetcore-app/

**Comment by KunalSaini** [comment]
Nicely explained,
However, how can I programmatically know what ports the application is launched with. I tried using the `IServerAddressesFeature` and parse the addresses in to C# [Uri](https://learn.microsoft.com/en-us/dotnet/api/system.uri?view=net-8.0) to be able to easily access port numbers. However, url's like `http://*:5003` can't be parsed by that class. 

Currently i am doing something like 
```C#
var portString = 
                Environment.GetEnvironmentVariable("ASPNETCORE_HTTP_PORTS") ??
                Environment.GetEnvironmentVariable("ASPNETCORE_HTTPS_PORTS") ??
                Environment.GetEnvironmentVariable("DOTNET_HTTP_PORTS") ??
                Environment.GetEnvironmentVariable("DOTNET_HTTPS_PORTS");
```
which seems brittle.

**Reply by andrewlock** [reply]
I have just the post for you 😄 https://andrewlock.net/how-to-automatically-choose-a-free-port-in-asp-net-core/#how-do-i-find-out-which-port-was-selected-

**Reply by KunalSaini** [reply]
Thanks Andrew, I am using the `IServerAddressFeature` based on your post, however I was trying to parse the `Addresses` into a C# Uri class. Which failed for Uri like `http://*:5003` as `*` is not a valid Uri character. I guess I can string parse. Something like, split the string at last index of `:` to get the port number.

**Reply by andrewlock** [reply]
I think it matters _when_ you're trying to grab them? In the post I point out that if you get them at the wrong time you get the invalid values, but later you get the real ones?

**Comment by rosdi** [comment]
I found a quite peculiar behavior. I am running an Aspire project, if I start the project in Visual Studio and selected 'https' profile, VS will start all the services using 'https' profile. In Rider though, it ***will always use the first profile configured*** as in `launchSettings.json`. 

For example, this launcSettings.json, in VS if I run the 'https' profile all child services under this Aspire project will be run using 'https' as well, but ***in Rider, 'http' will be used since it is the first profile***. Sounds like a bug in Rider to me, no? 

```json
{
  "$schema": "https://json.schemastore.org/launchsettings.json",
  "profiles": {
    "http": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "launchUrl": "swagger",
      "applicationUrl": "http://localhost:5372",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    },
    "https": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "launchUrl": "swagger",
      "applicationUrl": "https://localhost:7495;http://localhost:5372",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}

```

**Comment by Dave711** [comment]
Thank you so much for writing this awesome easy to read quick guide on how to configure dotnet to receive connections from outside the localhost!

**Comment by kkkristo** [comment]
Thank you for detailed explanation. I omitted https port related changes after .NET 6, and whenmy project upgraded to .NET 9, I experienced the app was listening to http port only. However my goal was to have different ports for developmnet and different in production. I prepared some kind of workaround like: 
```
if (!app.Environment.IsDevelopment())
{
    app.Urls.Add("http://localhost:5000");
    app.Urls.Add("https://localhost:5001");
```
it works but I'm somehow not happy with it. Also it affects webhosted deployment causing application error. Could you advise me some clean solution for setting up ports properly for local deployment, as web-deployed app apparently needs no extra configuration after upgrade to .NET 9.
